src/query/api/v1/handler/namespace/schema.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package namespace
+
+import (
+	"net/http"
+	"path"
+
+	clusterclient "github.com/m3db/m3/src/cluster/client"
+	"github.com/m3db/m3/src/cluster/kv"
+	"github.com/m3db/m3/src/dbnode/namespace/kvadmin"
+	"github.com/m3db/m3/src/query/api/v1/handler"
+	"github.com/m3db/m3/src/query/generated/proto/admin"
+	"github.com/m3db/m3/src/query/util/logging"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	"github.com/m3db/m3/src/x/net/http"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"go.uber.org/zap"
+)
+
+var (
+	// M3DBSchemaURL is the url for the M3DB schema handler.
+	M3DBSchemaURL = path.Join(handler.RoutePrefixV1, M3DBServiceSchemaPathName)
+
+	// SchemaHTTPMethod is the HTTP method used with this resource.
+	SchemaHTTPMethod = http.MethodPost
+)
+
+// SchemaHandler is the handler for namespace schema upserts.
+type SchemaHandler Handler
+
+// For unit test purpose.
+var newAdminService = kvadmin.NewAdminService
+
+// NewSchemaHandler returns a new instance of SchemaHandler.
+func NewSchemaHandler(client clusterclient.Client) *SchemaHandler {
+	return &SchemaHandler{client: client}
+}
+
+func (h *SchemaHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := logging.WithContext(ctx)
+
+	md, rErr := h.parseRequest(r)
+	if rErr != nil {
+		logger.Error("unable to parse request", zap.Error(rErr))
+		xhttp.Error(w, rErr.Inner(), rErr.Code())
+		return
+	}
+
+	opts := handler.NewServiceOptions("kv", r.Header, nil)
+	resp, err := h.Add(md, opts)
```

Comment:
Cant wait to rip that code out and put it in your service layer

Reply:
agree, so this handler just handle req/resp/err presentation.